### PR TITLE
Add initial common microgrid definitions & boilerplate

### PR DESCRIPTION
**Pull request by stefan-brus-frequenz**

### CODE REVIEW

File: proto/frequenz/api/common/component_category.proto [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  // A crypto miner.
+  COMPONENT_CATEGORY_CRYPTO_MINER = 8;
+
+  // An electrolyzer for converting water into hydrogen and oxygen.
+  COMPONENT_CATEGORY_ELECTROLYZER = 9;
+
+  // A heat and power combustion plant (CHP stands for combined heat and power).
+  COMPONENT_CATEGORY_CHP = 10;
+  }
```
Comment (tiyash-basu-frequenz) [inline proto/frequenz/api/common/component_category.proto]: The indentation looks wrong here.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        os:
+          - ubuntu-20.04
+        python-version:
+          - "3.11"
+    runs-on: ${{ matrix.os }}
+    steps:
+      - name: Fetch sources
+        uses: actions/checkout@v3
+        with:
+          submodules: true
```
Comment (tiyash-basu-frequenz) [inline .github/workflows/ci.yaml:48]: Would be nice if we add protolint here.

File: proto/frequenz/api/common/common.proto [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+// License:
+// MIT
+
+syntax = "proto3";
+
+package frequenz.api.common.common;
+
+// A set of lower and upper bounds for any metric.
+// The units of the bounds are always the same as the related metric.
+message Bounds {
```
Comment (tiyash-basu-frequenz) [inline proto/frequenz/api/common/common.proto]: We should remove everything except `DC` and `AC`, and rename this file to `electrical.proto`. Everything else is specific to the Microgrid API.

File: proto/frequenz/api/common/battery.proto [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+// MIT
+
+syntax = "proto3";
+
+package frequenz.api.common.battery;
+
+import "frequenz/api/common/common.proto";
+
+// Enumerated battery types.
+enum Type {
```
Comment (tiyash-basu-frequenz) [inline proto/frequenz/api/common/battery.proto]: We should just add component types here, as an extension to component categories.
So all this file should have is battery type enum, inverter type enum, ev charger type enum, etc.

Everything else like states, error codes, etc. are specific to the microgrid API.

File: proto/frequenz/api/common/battery.proto [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+  // Unspecified.
+  TYPE_UNSPECIFIED = 0;
+
+  // Li-ion batteries.
+  TYPE_LI_ION = 1;
+
+  // Lithium-Iron-Phosphate batteries.
+  TYPE_LI_FE_PO = 2;
```
Comment (tiyash-basu-frequenz) [inline proto/frequenz/api/common/battery.proto]: We should remove `LI_FE_PO` from here, because it a kind on Li-ion battery.

**Review by tiyash-basu-frequenz — CHANGES_REQUESTED** [review]

File: proto/frequenz/api/common/battery.proto (hunk 4 above)
Comment (tiyash-basu-frequenz) [reply proto/frequenz/api/common/battery.proto]: So maybe we could all all the types here, like `BatteryType`, `InverterType`, etc. And add these enums to `component_category.proto`, and rename it to `components.proto`.

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -8,7 +8,31 @@ concurrency:
...
+
+    steps:
+      - name: Fetch sources
+        uses: actions/checkout@v3
+        with:
+          submodules: true
+
+      - name: Get protolint
+        run: |
+          # checksum taken from: https://github.com/yoheimuta/protolint/releases/download/v0.38.1/checksums.txt
```
Comment (tiyash-basu-frequenz) [inline .github/workflows/ci.yaml]: Looks like [v0.44.0](https://github.com/yoheimuta/protolint/releases/tag/v0.44.0) is out.

File: proto/frequenz/api/common/components.proto [hunk 7]
```diff
@@ -46,3 +46,30 @@ enum ComponentCategory {
...
+
+  // Battery inverter.
+  INVERTER_TYPE_BATTERY = 1;
+
+  // Solar inverter.
+  INVERTER_TYPE_SOLAR = 2;
+
+  // Hybrid inverter.
+  INVERTER_TYPE_HYBRID = 3;
+}
```
Comment (tiyash-basu-frequenz) [inline proto/frequenz/api/common/components.proto:75]: I'd suggest adding EV charger type as well.

File: .github/workflows/ci.yaml (hunk 6 above)
Comment (stefan-brus-frequenz) [reply .github/workflows/ci.yaml]: Yeah... Looks like the checksums provided for this release are wrong though :facepalm: I will try another version

Comment (stefan-brus-frequenz) [reply .github/workflows/ci.yaml]: Nevermind, forgot to change a thing

**Comment by stefan-brus-frequenz** [comment]
The inverter "ev charger" type should go, that was a mistake. The enum types for sensor and ev charger from the microgrid API should be added as separate enums.

**Review by tiyash-basu-frequenz — DISMISSED** [review]
LGTM! Thanks for taking care of this.

**Review by tiyash-basu-frequenz — APPROVED** [review]
LGTM!

File: proto/frequenz/api/common/components.proto [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+//
+// Copyright:
+// Copyright 2023 Frequenz Energy-as-a-Service GmbH
+//
+// License:
+// MIT
+
+syntax = "proto3";
+
+package frequenz.api.common.component_category;
```
Comment (leandro-lucarella-frequenz) [inline proto/frequenz/api/common/components.proto:11]: This should be updated to `package frequenz.api.common.components;`, right?

File: proto/frequenz/api/common/components.proto [hunk 9]
```diff
@@ -0,0 +1,117 @@
+// Frequenz microgrid component categories.
```
Comment (leandro-lucarella-frequenz) [inline proto/frequenz/api/common/components.proto:1]: This should be updated to something like `// Frequenz microgrid components definitions`, right?

File: proto/frequenz/api/common/electrical.proto [hunk 10]
```diff
@@ -0,0 +1,199 @@
+// Contains definitions for electrical metrics (AC and DC).
```
Comment (leandro-lucarella-frequenz) [inline proto/frequenz/api/common/electrical.proto:1]: Maybe name the package `electrical_metrics` to keep consistency of names (the other are nouns and this one is an adjective). Or maybe is it worth to even nest it if we think at some point we could have other metrics definitions: `frequenz.api.common.metrics.electrical`.

**Review by leandro-lucarella-frequenz — COMMENTED** [review]
Sorry for arriving late to the party @stefan-brus-frequenz and @tiyash-basu-frequenz. In any case I have only very small comments, 2 of which should be trivial fixes and one that we could discuss if you think it has any merits :)

File: proto/frequenz/api/common/electrical.proto (hunk 10 above)
Comment (tiyash-basu-frequenz) [reply proto/frequenz/api/common/electrical.proto:1]: Makes sense.

Comment (leandro-lucarella-frequenz) [reply proto/frequenz/api/common/electrical.proto:1]: https://github.com/frequenz-floss/frequenz-api-common/pull/4

**Pull request merged**